### PR TITLE
Bound size of `TransactionPool` in memory

### LINKED ISSUE #3284: Bound size of `TransactionPool` in memory
Presently transactions are simply stored in a `BTreeMap` in memory, however, if we receive too many transactions too quickly, then this could grow over time, especially if the transactions are large.

### PR DESCRIPTION
**Pull request by nikurt**
Restrict the maximal size of transaction pools in validator nodes. Node that validator nodes have separate transaction pools for each tracked shard.
Transaction pool size can now be configured in `config.json` as a field `max_transaction_pool_size`.

Tested by running a validator and a non-validator node in betanet.

Fixes #3284

### CODE REVIEW

File: chain/pool/Cargo.toml [hunk 1]
```diff
@@ -12,7 +12,9 @@ edition = "2021"
 borsh = "0.9"
 once_cell = "1.5.2"
 rand = "0.7"
+tracing = "0.1"
 
+near-client-primitives = { path = "../client-primitives" }
 near-crypto = { path = "../../core/crypto" }
 near-metrics = { path = "../../core/metrics" }
-near-primitives = { path = "../../core/primitives" }
+near-primitives = { path = "../../core/primitives" }
```
Comment (mina86) [inline chain/pool/Cargo.toml:20]: Missing EOL and unnecessary diff.

File: chain/client/src/metrics.rs [hunk 2]
```diff
@@ -189,6 +189,10 @@ pub(crate) static TRANSACTION_RECEIVED_VALIDATOR: Lazy<IntGauge> = Lazy::new(||
     try_create_int_gauge("near_transaction_received_validator", "Validator received a transaction")
         .unwrap()
 });
+pub(crate) static TRANSACTION_DROPPED: Lazy<IntGauge> = Lazy::new(|| {
+    try_create_int_gauge("near_transaction_dropped", "Validator received and dropped a transaction")
```
Comment (mina86) [inline chain/client/src/metrics.rs:193]: Should we call it `near_transaction_dropped_validator` to match the other metric?

File: chain/client/src/metrics.rs [hunk 3]
```diff
@@ -189,6 +189,10 @@ pub(crate) static TRANSACTION_RECEIVED_VALIDATOR: Lazy<IntGauge> = Lazy::new(||
     try_create_int_gauge("near_transaction_received_validator", "Validator received a transaction")
```
Comment (mina86) [inline chain/client/src/metrics.rs:189]: The description is no longer accurate. Should be ‘Validator received and recorded a transaction’

File: chain/pool/src/lib.rs [hunk 4]
```diff
@@ -105,8 +115,14 @@ impl TransactionPool {
 
     /// Reintroduce transactions back during the chain reorg
     pub fn reintroduce_transactions(&mut self, transactions: Vec<SignedTransaction>) {
+        let mut dropped_transactions = 0;
         for tx in transactions {
-            self.insert_transaction(tx);
+            if self.insert_transaction(tx).is_err() {
+                dropped_transactions += 1;
+            }
+        }
```
Comment (mina86) [inline chain/pool/src/lib.rs:123]: Or if you want to be a hax0r, the following should work:
```rust
let dropped = transactions.into_iter().map(|tx| self.insert_transaction(tx)).filter(Result::is_err).count();
```

File: chain/pool/src/lib.rs [hunk 5]
```diff
@@ -253,11 +271,11 @@ mod tests {
         mut transactions: Vec<SignedTransaction>,
         expected_weight: u32,
     ) -> (Vec<u64>, TransactionPool) {
-        let mut pool = TransactionPool::new(TEST_SEED);
+        let mut pool = TransactionPool::new(TEST_SEED, TRANSACTION_POOL_SIZE);
         let mut rng = thread_rng();
         transactions.shuffle(&mut rng);
         for tx in transactions {
-            pool.insert_transaction(tx);
+            assert!(pool.insert_transaction(tx).is_ok());
```
Comment (mina86) [inline chain/pool/src/lib.rs:278]: It’s probably better to `unwrap`. While at the moment the method can fail for only one reason, the benefit of `unwrap` (apart from being shorter to write) is that it will output the error value if the result is not OK.

Applies below as well.

File: chain/pool/src/lib.rs [hunk 6]
```diff
@@ -49,10 +53,16 @@ impl TransactionPool {
...
+        signed_transaction: SignedTransaction,
+    ) -> Result<bool, Error> {
+        if self.unique_transactions.len() >= self.max_unique_transactions {
+            return Err(Error::TransactionPoolFull);
+        }
         if !self.unique_transactions.insert(signed_transaction.get_hash()) {
             // The hash of this transaction was already seen, skip it.
-            return false;
+            return Ok(false);
         }
```
Comment (mina86) [inline chain/pool/src/lib.rs:66]: Those two checks should be switched. Or rather, if we reached the size limit we should check if the transaction already exists and then return `Ok(false)`.

File: chain/pool/src/lib.rs [hunk 7]
```diff
@@ -465,4 +483,26 @@ mod tests {
...
+                Ok(_) => {
+                    num_inserted_tx += 1;
+                }
+                Err(err) => {
+                    assert!(matches!(err, TransactionPoolFull));
+                }
+            }
+        }
+        assert_eq!(num_inserted_tx, pool_size);
+    }
```
Comment (mina86) [inline chain/pool/src/lib.rs:507]:
```suggestion
    fn test_pool_size_restriction() {
        const POOL_SIZE: usize = 9;
        const OVER_LIMIT: usize = 3;

        let transactions =
            generate_transactions("alice.near", "alice.near", 1, (POOL_SIZE + OVER_LIMIT) as u64);

        let mut pool = TransactionPool::new(TEST_SEED, POOL_SIZE);

        for want_inserted in [POOL_SIZE, 0] {
            let mut inserted = 0;
            let mut duplicates = 0;
            let mut errors = 0;
            for tx in transactions.iter() {
                match pool.insert_transaction(tx.clone()) {
                    Ok(true) => inserted += 1,
                    Ok(false) => duplicates += 1,
                    Err(TransactionPoolFull) => errors += 1,
                    Err(err) => panic!("unexpected error: {}", err),
                }
            }
            let want = (want_inserted, POOL_SIZE - want_inserted, OVER_LIMIT);
            assert_eq!(want, (inserted, duplicates, errors));
        }
    }
```

**Review by mina86 — APPROVED** [review]
Perhaps describe a bit more in commit message what is the effect of the change?

File: chain/pool/src/lib.rs (hunk 4 above)
Comment (nikurt) [reply chain/pool/src/lib.rs:127]: I can, but I don't want to :)

**Comment by nikurt** [comment]
> Perhaps describe a bit more in commit message what is the effect of the change?

Done, please take another look.

**Review by mina86 — APPROVED** [review]
Code LGTM and I think from protocol side it should work as well though I’m less knowledgeable in that part.

File: chain/client/src/metrics.rs [hunk 8]
```diff
@@ -186,13 +186,23 @@ static NODE_BUILD_INFO: Lazy<IntCounterVec> = Lazy::new(|| {
...
 pub(crate) static TRANSACTION_RECEIVED_VALIDATOR: Lazy<IntGauge> = Lazy::new(|| {
-    try_create_int_gauge("near_transaction_received_validator", "Validator received a transaction")
-        .unwrap()
+    try_create_int_gauge(
+        "near_transaction_received_validator",
+        "Validator received and recorded a transaction",
+    )
+    .unwrap()
+});
+pub(crate) static TRANSACTION_DROPPED_VALIDATOR: Lazy<IntGauge> = Lazy::new(|| {
```
Comment (mzhangmzz) [inline chain/client/src/metrics.rs:195]: Should we add another metric for transaction pool size?

File: chain/client-primitives/src/types.rs [hunk 9]
```diff
@@ -33,6 +33,7 @@ pub enum Error {
     Chunk(near_chunks_primitives::Error),
     BlockProducer(String),
     ChunkProducer(String),
+    TransactionPoolFull,
```
Comment (mzhangmzz) [inline chain/client-primitives/src/types.rs:36]: I think this error should be added to near_chain_primitives::Error, since the TransactionPool struct sits in chain (whether that should be changed is another question). This will eliminate the need to add a dependency to client-pritimitives in chain, which feels a little weird.

**Comment by akhi3030** [comment]
@nikurt: this PR has not seen any activity in a long time.  I am assuming is no longer being actively developed.  I will close it.  Please reopen when you are ready to work on it again.
